permissions/rules/za.js

```diff
@@ -0,0 +1,129 @@
+const CanCan = require('cancan');
+const { isCounselorWhoCreated, isSupervisor, isCaseOpen } = require('./helpers');
+const Actions = require('../actions');
+const User = require('../user');
+const models = require('../../models');
+
+const cancan = new CanCan();
+const { can, allow } = cancan;
+
+const { Case } = models;
+
+allow(
```

Comment:
The CanCan documentation on NPM suggests that we could use the `typeof` the object as a way to define permissions:
```
// allow users to view all public posts
allow(User, 'view', Post, {public: true});
 
// allow users to edit and delete their posts
allow(User, ['edit', 'delete'], Post, (user, post) => post.authorId === user.id);
 
// allow editors to do anything with all posts
allow(Editor, 'manage', Post);
 
// allow admins to do anything with everything
allow(AdminUser, 'manage', 'all');
```

So if instead of defining everyone as a `User`, we could define `User`s, `Supervisor`s and `Admin`s and then do things like:
```
  allow(User, 'edit', Case, (user, case...) => ... isCaseOpen(caseObject) && isCounselorWhoCreated(...));

  allow(Supervisor, 'edit', Case, (user, case...) => ....includes(CLOSE_CASE));
```

And perhaps we could apply the same method for the cases, like:
```
  allow(User, 'edit', OpenCase, ...);
```

It may also be possible to define a class inheritance hierarchy where `User` is the parent of `Supervisor` is the parent of `Admin`, which would mean that children in the hierarchy inherit permissions.

None of this necessarily needs to be done now, but it seems like there may be a lot of opportunity to make this system more concise in the future.

Reply:
That's good to know, but I won't do this change now since it may make the next step more difficult: getting the permissions from a JSON. I won't get in details about it now, but I'm happy to discuss it later.